contract/r/gnoswap/gov/governance/execute_test.gno

```diff
@@ -367,3 +368,103 @@ func TestParameterRegistry2(t *testing.T) {
 		})
 	}
 }
+
+func TestAtomicExecution(t *testing.T) {
+	tests := []struct {
+		name        string
+		msgs        []string
+		setupMock   func(*ParameterRegistry)
+		expectError bool
+		validate    func(*testing.T, *ProposalInfo)
+	}{
+		{
+			name: "Should rollback state when second message fails",
+			msgs: []string{
+				"gno.land/r/gnoswap/v1/gns*EXE*SetAvgBlockTimeInMs*EXE*123",
+				"gno.land/r/gnoswap/v1/community_pool*EXE*TransferToken*EXE*gno.land/r/gnoswap/v1/gns,g17290cwvmrapvp869xfnhhawa8sm9edpufzat7d,905",
+			},
+			setupMock: func(r *ParameterRegistry) {
+				// 1st message should succeed
+				r.Register("gno.land/r/gnoswap/v1/gns", "SetAvgBlockTimeInMs", func(params []string) error {
+					return nil
+				})
+				// 2nd message should fail
+				r.Register("gno.land/r/gnoswap/v1/community_pool", "TransferToken", func(params []string) error {
+					return errors.New("insufficient balance")
+				})
+			},
+			expectError: true,
+			validate: func(t *testing.T, proposal *ProposalInfo) {
+				// state should be rolled back
+				uassert.False(t, proposal.State.Executed)
+				uassert.True(t, proposal.State.Passed) // voting should still be passed
+			},
+		},
+		{
+			name: "Should execute all messages atomically when all succeed",
+			msgs: []string{
+				"gno.land/r/gnoswap/v1/gns*EXE*SetAvgBlockTimeInMs*EXE*123",
+				"gno.land/r/gnoswap/v1/community_pool*EXE*TransferToken*EXE*gno.land/r/gnoswap/v1/gns,g17290cwvmrapvp869xfnhhawa8sm9edpufzat7d,905",
+			},
+			setupMock: func(r *ParameterRegistry) {
+				// all messages should succeed
+				r.Register("gno.land/r/gnoswap/v1/gns", "SetAvgBlockTimeInMs", func(params []string) error {
+					return nil
+				})
+				r.Register("gno.land/r/gnoswap/v1/community_pool", "TransferToken", func(params []string) error {
+					return nil
+				})
+			},
+			expectError: false,
+			validate: func(t *testing.T, proposal *ProposalInfo) {
+				// all messages should succeed
+				uassert.True(t, proposal.State.Executed)
+				uassert.True(t, proposal.State.Passed)
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// create a test proposal
+			proposal := &ProposalInfo{
```

Comment:
I think it needs to be executed by `executeProposal`. This proposal is not actually handled by handler.

Reply:
https://github.com/gnoswap-labs/gnoswap/pull/617/commits/b75c2991e959d78e7bcc48d3f66e2d4371138cf2